lib/active_merchant/billing/gateways/hi_pay.rb

```diff
@@ -0,0 +1,203 @@
+        post[:currency] = (options[:currency] || currency(money))
+        post[:amount] = money
+      end
+
+      def add_credit_card(post, credit_card)
+        post[:card_number] = credit_card.number
+        post[:card_expiry_month] = credit_card.month
+        post[:card_expiry_year] = credit_card.year
+        post[:card_holder] = credit_card.name
+        post[:cvc] = credit_card.verification_value
+      end
+
+      def add_address(post, options)
+        if options[:billing_address]
+          post[:streetaddress] = options[:billing_address][:address1] if options[:billing_address][:address1]
+          post[:streetaddress2] = options[:billing_address][:address2] if options[:billing_address][:address2]
+          post[:city] = options[:billing_address][:city] if options[:billing_address][:city]
+          post[:recipient_info] = options[:billing_address][:company] if options[:billing_address][:company]
+          post[:state] = options[:billing_address][:state] if options[:billing_address][:state]
+          post[:country] = options[:billing_address][:country] if options[:billing_address][:country]
+          post[:zipcode] = options[:billing_address][:zip] if options[:billing_address][:zip]
+          post[:country] = options[:billing_address][:country] if options[:billing_address][:country]
+          post[:phone] = options[:billing_address][:phone] if options[:billing_address][:phone]
+        end
+      end
+
+      def tokenize(payment_method, options = {})
+        post = {}
+        add_credit_card(post, payment_method)
+        post[:multi_use] = options[:multiuse] ? '1' : '0'
+        post[:generate_request_id] = '0'
+        commit('store', post, options)
+      end
+
+      def parse(body)
+        return {} if body.blank?
+
+        JSON.parse(body)
+      end
+
+      def commit(action, post, options = {})
+        begin
+          raw_response = ssl_post(url(action, options), post_data(post), request_headers)
+          response = parse(raw_response)
+        rescue ResponseError => e
+          raw_response = e.response.body
+          response = parse(raw_response)
+        end
+
+        Response.new(
+          success_from(action, response),
+          message_from(action, response),
+          response,
+          authorization: authorization_from(action, response),
+          test: test?
+        )
+      end
+
+      def success_from(action, response)
+        case action
+        when 'order'
+          response['state'] == 'completed'
+        when 'capture'
+          response['status'] == '118' && response['message'] == 'Captured'
+        when 'store'
+          response.include? 'token'
+        else
+          false
+        end
+      end
+
+      def message_from(action, response)
+        case action
+        when 'order', 'capture'
+          response['message']
+        end
+      end
+
+      def authorization_from(action, response)
+        case action
+        when 'order', 'capture'
```

Comment:
💬 Commit

As Nick commented on [Orbital PR](https://github.com/activemerchant/active_merchant/pull/4959#discussion_r1409550402), the expected way to build authorization strings is to always include the separators to have fixed structure on the reference string.

Reply:
Thanks I'll use it